waspc/data/Generator/templates/server/src/auth/providers/oauth/state.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  Response as ExpressResponse,
+  Request as ExpressRequest,
+} from "express";
+import { generateCodeVerifier, generateState } from "arctic";
+
+import type { ProviderConfig } from "wasp/auth/providers/types";
+
+import { setOAuthCookieValue, getOAuthCookieValue } from "./cookies.js";
+
+export type StateType = 'state' | 'codeVerifier';
+
+export function generateAndStoreOAuthState<ST extends StateType>(
+  stateTypes: ST[],
+  provider: ProviderConfig,
+  res: ExpressResponse,
+): { [name in ST]: string } {
+  const result = {} as { [name in StateType]: string }
+
+  if (stateTypes.includes('state' as ST)) {
+    const state = generateState();
+    setOAuthCookieValue(provider, res, 'state', state);
+    result.state = state;
+  }
+
+  if (stateTypes.includes('codeVerifier' as ST)) {
+    const codeVerifier = generateCodeVerifier();
+    setOAuthCookieValue(provider, res, 'codeVerifier', codeVerifier);
+    result.codeVerifier = codeVerifier;
+  }
+
+  return result as { [name in ST]: string };
```

Comment:
A less error prone way to implement this is by generating the resulting state first, and then using that to set the value. In pseudocode:
```typescript
// construct the result
setOAuthCookieValue(provider, res, result); // this function would iterate the object and set all cookies
return result
```
I probably wouldn't do this now though and would just merge this. Not a big deal.